pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -95,6 +113,12 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
         exclude: (Node) -> Boolean = { false } // exclude none by default
     ) = node.getChildren().filterNot(exclude).forEach { build(it, parent) }
 
+    /**
+     * Builds [node] using its default building behavior. Used to build a child of a [Node] in place
+     * of itself.
+     */
+    private fun buildInstead(node: Node?) = build(node, parent = null)
```

Comment:
it's not clear to my why this method is called "build instead"

Reply:
Yes, this is the main part I wanted feedback on.

During development, I saw a pattern where a node would need to delegate its conversion to one of its children. For example, a `MethodDeclaration` is usually turned into an `ASTMethod` within `buildMethodDeclaration`, but in one case the body of the method (`CompoundStatement` -> `ASTBlockStatement`) should be built in its place.

Originally I expected this to be handled by calling the regular `build` function, e.g.
```kotlin
when {
  node.isAnonymousInitializationCode() && !node.hasKeyword(Keyword.STATIC) -> build(node.body, parent)
  // ...
}
```
but this resulted in an issue where the node would be added twice to its parent, for the following reason:
1. The `ASTBlockStatement` node is returned by `build(node.body, parent)`, is returned as the result of `buildMethodDeclaration`, is assigned to `wrapper` in `build` and is added to `parent`.
2. The `build(node.body, parent)` call that constructs the `ASTBlockStatement` assigns it to its own `wrapper` and also adds it to the same `parent`.

Either of these cases can be disabled to fix the duplication issue, so I set `parent` to `null` in `buildInstead` to disable the latter case. I wanted to provide a name for this operation, so I extracted it into a function. I have no preference for the name as long as it makes the purpose clear.

P.S. I definitely see this design as flawed (that the node is both returned by `build` and `setParent` is called by `build`), but I haven't found an alternative that works better yet. I think this would be a good point for me to continue looking for more alternatives while I haven't added many new nodes.

Reply:
I found another way to factor the building functions that I am more satisfied with. See [`328887f`](https://github.com/eklimo/pmd/pull/5/commits/328887f8879afc416cc82720711f0d369678a6e8). Now `setParent` has to be called intentionally and `parent` doesn't have to be set to `null` confusingly.